Support dumping seccomp on aarch64

saagarjha opened the pull request:
This adds support for dumping seccomp BPF on aarch64–doing it by hand for DEF CON is fine once, but I'd rather not have to do it again :) The setup I used to test this is extraordinarily janky (CTF challenge-provided binary/kernel running in QEMU using initramfs+a homegrown rootfs that has literally every single dependency in it available offline, since I can't figure out how to set up networking). Seems to work on some simple binaries I tested it against, but I couldn't figure out how to get the actual test suite running so if you have guidance there it would be much appreciated.

david942j commented:
Thanks for this contribution! It basically looks fine but the tests failed with unclear reasons (I would guess it's related to the ptrace change)

Also please add tests to ensure the coverage reaches 100% ;)

david942j commented on ext/ptrace/ptrace.c:
```diff
@@ -21,7 +23,18 @@ ptrace_peekdata(VALUE _mod, VALUE pid, VALUE addr, VALUE _data) {
+  size_t offset = NUM2LONG(off);
+  long val;
+  size_t size = offset + sizeof(val);
+  char *regs = malloc(size);
```
check the return value is not NULL

david942j replied:
Or maybe directly declare `struct user_pt_regs regs`

saagarjha replied:
Yeah so the code here is kind of strange, perhaps I should have added a comment explaining why I'm not using `struct user_pt_regs`. The TL;DR is that I couldn't find a header that declared it consistently: my x86_64 Ubuntu machine has `struct user_regs_struct` declared in /usr/include/x86_64-linux-gnu/sys/user.h, but the aarch64 Alpine (musl-based) rootfs I was using has `struct user_pt_regs` in /usr/include/asm/ptrace.h. The untyped, dynamically-sized buffer is my way of getting around this by not using that struct at all, although it has the safety issues you've outlined. I'm open to suggestions as to how you think I should do this :)

david942j replied:
Oh I didn't notice this. Maybe we can check the source of GDB to see how it handles this issue, though I guess it simply uses if-def guards to perform arch-dependent operations.

Anyways, in this case it's fine to use malloc/free here, but please add a comment on why we can't use `struct user_regs_struct` or similar structures here.

And for the validation of `offset`, set a reasonable bound (4096) is enough (but also add a comment why we can't compare it with `sizeof(...)`). I just don't want an invalid input from Ruby could lead the program crash.

david942j commented on ext/ptrace/ptrace.c:
```diff
@@ -21,7 +23,18 @@ ptrace_peekdata(VALUE _mod, VALUE pid, VALUE addr, VALUE _data) {
+  // PTRACE_GETREGSET which is fairly similar. Since i386 and x86_64 support it
+  // too, just use that unconditionally to present the same API for the
+  // registers we care about.
+  size_t offset = NUM2LONG(off);
```
Can we check  this value is valid? Something like it should be less than `sizeof(struct user_pt_regs)`

saagarjha replied:
Ideally yes, but see above.

saagarjha commented on ext/ptrace/ptrace.c:
```diff
@@ -21,7 +23,18 @@ ptrace_peekdata(VALUE _mod, VALUE pid, VALUE addr, VALUE _data) {
+  size_t size = offset + sizeof(val);
+  char *regs = malloc(size);
+  struct iovec vec = { regs, size };
+  ptrace(PTRACE_GETREGSET, NUM2LONG(pid), NT_PRSTATUS, &vec);
```
Sure, I can do that. What should I return if this fails, though? -1?

david942j replied:
Yap `-1` should be good enough. Ideally returning the errno would be better, but that also needs the Ruby side to properly handle the error cases, which is not necessary to be done in this PR.

saagarjha commented:
Thanks for looking at this! A couple of questions above, but regarding testing: I'm not a Rubyist by any means so I'm mostly just stumbling around trying things while working with this. To test locally I have been running `bundle install` and then `bundle exec rake compile`, then running `bundle exec seccomp-tools dump` on some test binaries. Is this the intended way to work with the project? If not, how should I be doing so? If I run `bundle exec rake` on my x86_64 machine (which I think runs the tests?) it hangs so I'm not sure if that is the right way to do it.

Regarding adding tests: I tried this against a binary from DEF CON and [this example](https://gist.github.com/fntlnz/08ae20befb91befd9a53cd91cdc6d507) I found online, tweaked slightly. For the latter I changed out the architecture when installing the filter and changed it block `writev`, which is what musl uses for `printf`. For the former, I need to do a more roundabout technique because it forks a thread and applies the filter to that, so I grab the subtask ID from procfs and run seccomp-tools on a PID after launching it. Oh, and this is all running in the sketchiest qemu build ever–I'm not sure it is really fit for CI, if that's your goal. Again, your thoughts are very welcome :)

david942j commented:
Yes `bundle exec rake` on an x86_64 machine is the expect way to run tests. I don't know why it hangs to you though, but indeed the tests hang on travis CI as well (https://travis-ci.com/github/david942j/seccomp-tools/jobs/372610410).
[This](https://travis-ci.com/github/david942j/seccomp-tools/jobs/372360331) is the latest build on master branch (4 days ago) and it works fine, so I assume the hang is caused by your patches.

My point is you can still add tests against the functions such as `disasm` and `emu`.
Only the function `dump` needs the binary to be run, I can take care of that after this PR.

saagarjha commented:
Turns out Yama was being too mindful and rejecting trace requests :/ Then I fixed some of the offsets for i386, since Linux apparently uses a 4 byte pointer size with `PTRACE_GETREGSET` but native pointer size for `PTRACE_PEEKUSER`…

I've added comments and checks for the things you've mentioned in ptrace.c, and there's a test for disasm and emu based off of the bdooos challenge from DEF CON–the binary for which I've committed as well. Let me know if there's anything else you'd like me to add.

david942j commented on ext/ptrace/ptrace.c:
```diff
@@ -19,10 +22,49 @@ ptrace_peekdata(VALUE _mod, VALUE pid, VALUE addr, VALUE _data) {
+  struct iovec vec = { regs, size };
+  if (ptrace(PTRACE_GETREGSET, NUM2LONG(pid), NT_PRSTATUS, &vec) != -1 && vec.iov_len >= size) {
+    memcpy(&val, regs + offset, width);
+  } else {
```
missing `free(regs)` in the `else` block

david942j commented on spec/util_spec.rb:
```diff
@@ -14,7 +14,7 @@
     expect(described_class.system_arch).to be :i386
     RbConfig::CONFIG['host_cpu'] = 'aarch64'
-    expect(described_class.system_arch).to be :unknown
+    expect(described_class.system_arch).to be :aarch64
```
add another arch and test it should be `:unknown`

david942j commented on ext/ptrace/ptrace.c:
```diff
@@ -19,10 +22,49 @@ ptrace_peekdata(VALUE _mod, VALUE pid, VALUE addr, VALUE _data) {
+  if (width == sizeof(val.val32)) {
+    return LONG2NUM(val.val32);
+  } else if (width == sizeof(val.val64)) {
+    return LONG2NUM(val.val64);
```
fix the compiler warning shown in https://travis-ci.com/github/david942j/seccomp-tools/jobs/378197696

saagarjha commented:
Done, done, and done :)

david942j commented:
Merged, thanks for the contribution!
